File: Swift-Coordinator/Swift-Coordinator/Utils/NetworkingObjectivC/Request.m
//
//  Request.m
//  Swift-Coordinator
//
//  Created by Fabricio Pujol on 25/03/20.
//  Copyright © 2020 Luiz Felipe Trindade. All rights reserved.
//

#import "Request.h"

@implementation Request

//singleton
+(instancetype)sharedRequestManager{
    static Request *sharedRequestManager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedRequestManager = [[self alloc]init];
    });
    return sharedRequestManager;
}



- (void)fetchNowPlaying: (id<RequestProtocol>)delegate {
    NSLog(@"Fecthing Now Playing");
    NSString *urlString = @"https://api.themoviedb.org/3/movie/now_playing?api_key=ec133e4315b807c11f344060f8b97991&language=en-US&page=1";

    NSURL *url = [NSURL URLWithString:urlString];

    [[NSURLSession.sharedSession dataTaskWithURL:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {

        NSLog(@"finished Fecthing NowPlaying...");

        NSError *err;
        NSArray *nowPlayingJSON = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&err];

        if(err){
            NSLog(@"Failed to serialize into JSON: %@", err);
            return;
        }

        NSMutableArray <Movie *> *movies = NSMutableArray.new;

        for (NSDictionary *nowPlayingDict in [nowPlayingJSON valueForKey:@"results"]) {
            NSString *title = nowPlayingDict[@"original_title"];
            NSString *overview = nowPlayingDict[@"overview"];
            NSString *backdrop = nowPlayingDict[@"backdrop_path"];
            NSString *movieId = nowPlayingDict[@"id"];
            NSString *language = nowPlayingDict[@"original_language"];
            NSString *posterPath = nowPlayingDict[@"poster_path"];
            NSMutableArray *genres = nowPlayingDict[@"genre_ids"];
            NSNumber *popularity = nowPlayingDict[@"popularity"];
            NSNumber *voteAverage = nowPlayingDict[@"vote_average"];
            NSNumber *voteCount = nowPlayingDict[@"vote_count"];

            Movie *movie = Movie.new;
            movie.vote_count = voteCount;
            movie.vote_average = voteAverage;
            movie.popularity = popularity;
            movie.poster_path = posterPath;
            movie.original_language = language;
            movie.original_title = title;
            movie.overview = overview;
            movie.backdrop_path = backdrop;
            movie.pk_id = movieId;
            movie.genre_ids = genres;
            [movies addObject:movie];
        }

        [delegate didRetrieveResponse:movies forRequest:@"NowPlaying"];
    }] resume];
}

- (void)fetchPopular: (id<RequestProtocol>)delegate {
    NSLog(@"Fecthing Now Playing");
    NSString *urlString = @"https://api.themoviedb.org/3/movie/popular?api_key=ec133e4315b807c11f344060f8b97991&language=en-US&page=1";

    NSURL *url = [NSURL URLWithString:urlString];

    [[NSURLSession.sharedSession dataTaskWithURL:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {

        NSLog(@"finished Fecthing NowPlaying...");

        NSError *err;
        NSArray *nowPlayingJSON = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&err];

        if(err){
            NSLog(@"Failed to serialize into JSON: %@", err);
            return;
        }

        NSMutableArray <Movie *> *movies = NSMutableArray.new;

        for (NSDictionary *nowPlayingDict in [nowPlayingJSON valueForKey:@"results"]) {
            NSString *title = nowPlayingDict[@"original_title"];
            NSString *overview = nowPlayingDict[@"overview"];
            NSString *backdrop = nowPlayingDict[@"backdrop_path"];
            NSString *movieId = nowPlayingDict[@"id"];
            NSString *language = nowPlayingDict[@"original_language"];
            NSString *posterPath = nowPlayingDict[@"poster_path"];
            NSMutableArray *genres = nowPlayingDict[@"genre_ids"];
            NSNumber *popularity = nowPlayingDict[@"popularity"];
            NSNumber *voteAverage = nowPlayingDict[@"vote_average"];
            NSNumber *voteCount = nowPlayingDict[@"vote_count"];

            Movie *movie = Movie.new;
            movie.vote_count = voteCount;
            movie.vote_average = voteAverage;
            movie.popularity = popularity;
            movie.poster_path = posterPath;
            movie.original_language = language;
            movie.original_title = title;
            movie.overview = overview;
            movie.backdrop_path = backdrop;
            movie.pk_id = movieId;
            movie.genre_ids = genres;
            [movies addObject:movie];
        }
        [delegate didRetrieveResponse:movies forRequest:@"PopularMovies"];
    }] resume];
}

- (void)fetchGenres: (int) id request:(id<RequestProtocol>) delegate {
    NSLog(@"Fecthing Genres");
    NSString* idString = [@(id) stringValue];
    NSString *urlString = [NSString stringWithFormat:@"https://api.themoviedb.org/3/movie/%@?api_key=ec133e4315b807c11f344060f8b97991&language=en-US", idString];

    NSURL *url = [NSURL URLWithString:urlString];

    [[NSURLSession.sharedSession dataTaskWithURL:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {

        NSLog(@"finished Fecthing Genres");

        NSError *err;
        NSArray *nowPlayingJSON = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&err];

        if(err){
            NSLog(@"Failed to serialize into JSON: %@", err);
            return;
        }

        NSMutableArray <Genre *> *genres = NSMutableArray.new;

        for (NSDictionary *nowPlayingDict in [nowPlayingJSON valueForKey:@"genres"]) {
            NSString *name = nowPlayingDict[@"name"];
            NSNumber *id = nowPlayingDict[@"id"];

            Genre *genre = Genre.new;
            genre.name = name;
            genre.id = id;
            [genres addObject:genre];
        }

        NSLog(@"%@", genres);

        [delegate didRetrieveResponse:genres forRequest:@"MovieCategory"];
    }]
     resume];
}

@end

File: Swift-Coordinator/Swift-Coordinator/Model/Request.swift
//
//  Request.swift
//  Swift-Coordinator
//
//  Created by Fabricio Pujol on 16/08/19.
//  Copyright © 2019 Luiz Felipe Trindade. All rights reserved.
//

import Foundation

struct Request {
    let resourceURLNowPlaying: URL
    let resourceURLPopular: URL
    let resourceURLGenre: URL
    let apiKey = "ec133e4315b807c11f344060f8b97991"
    let region = "BR"
    
    init() {

        let resourceStringNowPlaying = "https://api.themoviedb.org/3/movie/now_playing?api_key=\(apiKey)&language=en-US&region=\(region)"
        guard let resourceURLNowPlaying = URL(string: resourceStringNowPlaying) else {fatalError("invalid UrlPath")}
        
        let resourceStringPopular = "https://api.themoviedb.org/3/movie/popular?api_key=\(apiKey)&language=en-US&page=1&region=\(region)"
        guard let resourceURLPopular = URL(string: resourceStringPopular) else {fatalError("invalid UrlPath")}
        
        let resourceStringGenre = "https://api.themoviedb.org/3/genre/movie/list?api_key=\(apiKey)&language=en-US"
        guard let resourceURLGenre = URL(string: resourceStringGenre) else {fatalError("invalid UrlPath")}
        
        print(resourceStringNowPlaying)
        print(resourceURLPopular)
        
        self.resourceURLNowPlaying = resourceURLNowPlaying
        self.resourceURLPopular = resourceURLPopular
        self.resourceURLGenre = resourceURLGenre
    }
    
    func fetchMovies(completitionHandler: @escaping (Data?) -> Void){
        let session = URLSession.shared
        
        let task = session.dataTask(with: resourceURLNowPlaying) { (data,response, error) in
            
            if let _ = error {
                completitionHandler(nil)
            } else {
                if let data = data {
                    completitionHandler(data)
                } else {
                    completitionHandler(nil)
                }
            }
        }
        task.resume()
    }
    
    func fetchPopularMovies(completitionHandler: @escaping (Data?) -> Void){
        let session = URLSession.shared
        
        let task = session.dataTask(with: resourceURLPopular) { (data,response, error) in
            
            if let _ = error {
                completitionHandler(nil)
            } else {
                if let data = data {
                    completitionHandler(data)
                } else {
                    completitionHandler(nil)
                }
            }
        }
        task.resume()
    }
    
    func fetchGenre(completitionHandler: @escaping (Data?) -> Void){
        let session = URLSession.shared
        let task = session.dataTask(with: resourceURLGenre) { (data,response, error) in
            
            if let _ = error {
                completitionHandler(nil)
            } else {
                if let data = data {
                    completitionHandler(data)
                } else {
                    completitionHandler(nil)
                }
            }
        }
        task.resume()
    }
}
